Improve Authorizer implementation

mcaskill opened the pull request:
The `Authorizer` now extends and interface, an abstract class, and a variety of new methods to check permissions against roles and resources.

Previously, the Authorizer provided a limited interface for interacting with the ACL and evaluating permissions against roles (one issue being the lack of support for multiple roles).

### Old API

- `rolesAllowed()` — Deprecated in favour of `anyRolesGrantedAll()`
- `userAllowed()` — Deprecated in favour of `anyRolesGrantedAll()` (via `isUserGranted()`)
- `resource` renamed to `defaultResource`

Internally, the old permission checking methods now use the new methods (which fixes support for multiple roles) but preserves the default behavior of allowing access to everything.

### New API

- `isRoleGrantedAll()` — Check if access is granted to the role, and the resource, for all permissions.
  - `allRolesGrantedAll()` — Check if access is granted to all roles, and the resource, for all permissions.
  - `anyRolesGrantedAll()` — Check if access is granted to any one of the roles, and the resource, for all permissions.
  - `isUserGranted()` — Check if access is granted to the user's role(s), and the resource, for permissions.
- `isRoleGrantedAny()` — Check if access is granted to the role, and the resource, for any one of the permissions.
  - `allRolesGrantedAny()` — Check if access is granted to all roles, and the resource, for any one of the permissions.
  - `anyRolesGrantedAny()` — Check if access is granted to any one of the roles, and the resource, for any one of the permissions.
- `isAllowed()` — Check if the role has access to the resource and privilege.
- `hasRole()` — Check if the role is registered.
- `inheritsRole()` — Check if the role inherits from another role.
- `hasResource()` — Check if the resource is registered.
- `inheritsResource()` — Check if the resource inherits from another resource.

### Example

#### Example #&#8203;1

Using the new API with the default "charcoal" resource.

```php
if (!$authorizer->isUserGranted($user, Authorizer::DEFAULT_RESOURCE, 'edit')) {
    return $response->withStatus(403);
}
```

#### Example #&#8203;2

```php
public function isAuthorizedToManageOthers()
{
    $obj        = $this->obj();
    $objType    = $obj->objType();
    $authorizer = $this->authorizer();

    if ($authorizer->hasResource($objType)) {
        $user = $this->authenticator()->getUser();
        if ($user) {
            return $authorizer->isUserGranted($user, $objType, 'object/manage/others');
        }
    }

    return false;
}

protected function prepareAuthorship(ModelInterface $obj)
{
    $old    = $this->prevObj;
    $userId = $this->authenticator()->getUserId();

    if ($old->hasAuthor($userId) && !$obj->hasAuthor($userId)) {
        // Redirect if current user is no longer an author
        if (!$this->isAuthorizedToManageOthers()) {
            $url = $this->getObjectBrowseUrl();
            $url = $obj->renderTemplate($url);
            $this->setSuccessUrl($url);
        }
    }
}
```

mcaskill committed:
[!] Improve Authorizer implementation

The Authorizer now supports and interface, a base class, and a variety of methods to check permissions against roles.

Deprecated:
- Methods `rolesAllowed()` and `userAllowed()` on `Authorizer` in favour of `AbstractAuthorizer::anyRolesGrantedAll()`

Added:
- Methods `AbstractAuthorizer::isRoleGrantedAll()` and `isRoleGrantedAny()` to test roles, resource, and permissions
- Method `AbstractAuthorizer::isUserGranted()` to test user, resource, and permissions
- Method `AbstractAuthorizer::isAllowed()` to test singular role, resource, and permission
- Methods `AbstractAuthorizer::hasRole()` and `AbstractAuthorizer::hasResource()` to lookup roles and resources

BeneRoch reviewed (approved):
Sounds good to me. Je pourrait tester avec Terrebonne si tout fonctionne puisque j'ai un système de rôle qui bloque l'accès à certains objets au niveau de la route et qui aurait pu bénéficier d'être une hiérarchie de rôles!

mcaskill commented:
This feature request was originally developed for USports which required support for tiered roles as well as multiple roles as well as privileges such as constraining editing of articles and allowing certain users to be able to edit articles of other users.

mcaskill committed:
Fix type-hinting on AuthAwareTrait

Amends: 8703142aebca046d408bff4b3bb41589b179dfb4

Fixed:
- Type-hint `AuthenticatorInterface` instead of `Authenticator`
